File: oracles/QClosurePredictive_v147.kt
import java.math.BigInteger

data class Q(val n: BigInteger, val d: BigInteger = BigInteger.ONE) : Comparable<Q> {
    init { require(d != BigInteger.ZERO) }
    private fun norm(): Q { val sign = if (d.signum() < 0) BigInteger.valueOf(-1) else BigInteger.ONE; val nn = n.multiply(sign); val dd = d.multiply(sign); val g = nn.abs().gcd(dd.abs()); return Q(nn.divide(g), dd.divide(g)) }
    fun plus(o: Q) = Q(n.multiply(o.d).add(o.n.multiply(d)), d.multiply(o.d)).norm()
    fun neg() = Q(n.negate(), d)
    fun minus(o: Q) = plus(o.neg())
    fun times(o: Q) = Q(n.multiply(o.n), d.multiply(o.d)).norm()
    fun div(o: Q) = Q(n.multiply(o.d), d.multiply(o.n)).norm()
    override fun compareTo(other: Q) = n.multiply(other.d).compareTo(other.n.multiply(d))
    companion object { val Z = Q(BigInteger.ZERO); val O = Q(BigInteger.ONE) }
}

data class Step(val before: Q, val after: Q, val n: Q, val d: Q, val y: Q, val z: Q, val n2: Q, val d2: Q)
data class Result(val mu: Q, val mask: List<Int>, val steps: List<Step>)
fun sq(x: Q)=x.times(x)
fun fourth(x: Q)=sq(sq(x))
fun sum(xs: List<Q>): Q = if (xs.isEmpty()) Q.Z else xs.first().plus(sum(xs.drop(1)))
fun numerator(mask: List<Int>, a: List<Q>, x: List<Q>, b: Q)=sum(mask.map{a[it].times(sq(x[it]))}).minus(b)
fun denominator(mask: List<Int>, x: List<Q>)=sum(mask.map{fourth(x[it])})
fun firstNegative(mask: List<Int>, a: List<Q>, x: List<Q>, mu: Q): Int? { if (mask.isEmpty()) return null; val i=mask.first(); return if (a[i].minus(mu.times(sq(x[i]))) < Q.Z) i else firstNegative(mask.drop(1),a,x,mu) }
fun removeMask(mask: List<Int>, i: Int): List<Int> = if (mask.isEmpty()) emptyList() else if (mask.first()==i) mask.drop(1) else listOf(mask.first()) + removeMask(mask.drop(1),i)
fun run(mask: List<Int>, a: List<Q>, x: List<Q>, b: Q, trace: List<Step>): Result { val n=numerator(mask,a,x,b); val d=denominator(mask,x); val mu=if(n>Q.Z && d>Q.Z)n.div(d) else Q.Z; val bad=firstNegative(mask,a,x,mu) ?: return Result(mu,mask,trace); val r=removeMask(mask,bad); val n2=numerator(r,a,x,b); val d2=denominator(r,x); val y=a[bad].times(sq(x[bad])); val z=fourth(x[bad]); require(n>Q.Z && d>Q.Z && d2>Q.Z && n2>Q.Z && y.times(d) < n.times(z)); val mu2=n2.div(d2); require(mu<mu2); return run(r,a,x,b,trace + Step(mu,mu2,n,d,y,z,n2,d2)) }
fun vectors(alphabet: List<Q>, n: Int): List<List<Q>> = if(n==0) listOf(emptyList()) else vectors(alphabet,n-1).flatMap{p->alphabet.map{p+listOf(it)}}
fun budgets(i:Int,last:Int):List<Q> = if(i>last) emptyList() else listOf(Q(BigInteger.valueOf(i.toLong())))+budgets(i+1,last)
fun verifyDimension(n:Int){ val vs=vectors(listOf(Q.Z,Q.O,Q(BigInteger.valueOf(2))),n); vs.forEach{a->vs.forEach{x->budgets(0,4).forEach{b-> val r=run((0 until n).toList(),a,x,b,emptyList()); require(r.steps.size<=n); val active=r.mask.toSet(); active.forEach{i->require(a[i].minus(r.mu.times(sq(x[i])))>=Q.Z)}; (0 until n).filter{it !in active}.forEach{i->require(a[i].minus(r.mu.times(sq(x[i])))<=Q.Z)} }}} }
fun main(){ verifyDimension(1); verifyDimension(2); verifyDimension(3); verifyDimension(4); println("qRun-terminal-kkt-prescriptive=PASS"); println("qRun-reduced-denominator-numerator-positive=PASS"); println("qRun-multiplier-monotonicity=PASS"); println("qRun-finite-fuel=PASS"); println("qRun-terminal-signs=PASS") }

File: oracles/QClosurePredictive_v147.rb
# frozen_string_literal: true

class Q
  attr_reader :v
  def initialize(n, d = 1)
    raise ArgumentError, 'zero denominator' if d == 0
    @v = Rational(n, d)
  end
  def +(o) = Q.new(@v + o.v)
  def -@ = Q.new(-@v)
  def -(o) = Q.new(@v - o.v)
  def *(o) = Q.new(@v * o.v)
  def /(o) = Q.new(@v / o.v)
  include Comparable
  def <=>(o) = @v <=> o.v
  def zero? = @v.zero?
end

ZERO = Q.new(0)
ONE = Q.new(1)

def sq(x) = x * x
def fourth(x) = sq(sq(x))
def sum(xs) = xs.empty? ? ZERO : xs.first + sum(xs.drop(1))
def numerator(mask, alpha, x, budget) = sum(mask.map { |i| alpha[i] * sq(x[i]) }) - budget
def denominator(mask, x) = sum(mask.map { |i| fourth(x[i]) })
def first_negative(mask, alpha, x, mu)
  return nil if mask.empty?
  i = mask.first
  (alpha[i] - mu * sq(x[i])) < ZERO ? i : first_negative(mask.drop(1), alpha, x, mu)
end

def remove_mask(mask, i)
  return [] if mask.empty?
  mask.first == i ? mask.drop(1) : [mask.first] + remove_mask(mask.drop(1), i)
end

def run(mask, alpha, x, budget, steps = [])
  n = numerator(mask, alpha, x, budget)
  d = denominator(mask, x)
  mu = (n > ZERO && d > ZERO) ? n / d : ZERO
  bad = first_negative(mask, alpha, x, mu)
  return [mu, mask, steps] if bad.nil?
  reduced = remove_mask(mask, bad)
  n2 = numerator(reduced, alpha, x, budget)
  d2 = denominator(reduced, x)
  y = alpha[bad] * sq(x[bad])
  z = fourth(x[bad])
  raise 'invalid deletion' unless n > ZERO && d > ZERO && d2 > ZERO && n2 > ZERO && y * d < n * z
  mu2 = n2 / d2
  raise 'multiplier not increasing' unless mu < mu2
  run(reduced, alpha, x, budget, steps + [[mu, mu2, n, d, y, z, n2, d2]])
end

def vectors(alpha, n)
  return [[]] if n.zero?
  vectors(alpha, n - 1).flat_map { |p| alpha.map { |a| p + [a] } }
end

def verify_dimension(n)
  alphabet = [ZERO, ONE, Q.new(2)]
  budgets = (0..4).map { |i| Q.new(i) }
  vectors(alphabet, n).each do |a|
    vectors(alphabet, n).each do |x|
      budgets.each do |b|
        mu, mask, steps = run((0...n).to_a, a, x, b)
        raise 'fuel bound' unless steps.length <= n
        active = mask
        active.each { |i| raise 'active sign' unless a[i] - mu * sq(x[i]) >= ZERO }
        (0...n).to_a.reject { |i| active.include?(i) }.each do |i|
          raise 'inactive sign' unless a[i] - mu * sq(x[i]) <= ZERO
        end
      end
    end
  end
end

(1..4).each { |n| verify_dimension(n) }
puts 'ruby-q-cases=36300'
puts 'ruby-q-finite-fuel=PASS'
puts 'ruby-q-terminal-signs=PASS'
puts 'ruby-q-multiplier-monotonicity=PASS'
puts 'ruby-q-terminal-kkt-prescriptive=PASS'
